### ISSUE TITLE
Constraint specification for SMARTY XML ForceField

### ISSUE THREAD

**Issue by jchodera** [open-forcefield-group/smarty#100]
At some point, we'll need to expand `ForceField` to allow us to specify which bonds or angles should be constrained.

One way to do this is to extend the `ffxml` format to allow this to be specified via SMIRKS. For example:

``` XML
<Constraints>
  <!-- bonds involving hydrogen -->
  <Constraint smirks="[#1:1]-[*:2]"/> <!-- add constraint between atoms 1 and 2 using equilibrium bond distance -->
  <!-- angles involving hydrogen -->
  <Constraint smirks="[#1:1]-[*:2]-[#1:3]"/> <!-- add constraint between atoms 1 and 3 using auto-calculated distance from equilibrium bond and angles -->
</Constraints>
```

**Comment by davidlmobley** [comment]
This seems reasonable to me. I can discuss with Chris on Monday if you want. I'll tag him here but he's probably out for the weekend. ( @cbayly13 ?)

**Comment by davidlmobley** [comment]
@jchodera - this is where we stand on constraints.

**Comment by jchodera** [comment]
OK! This is super easy to implement if we get the thumbs-up.

We might want to also draft an example for rigid waters. Would that be:
```XML
<Constraints>
  <!-- constrain water O-H bond to equilibrium bond length -->
  <Constraint smirks="[#1X1:1]-[#6X2:2]-[#1X1]"/> 
  <!-- constrain water H...H, calculating equilibrium length from H-O-H equilibrium angle and H-O equilibrium bond lengths -->
  <Constraint smirks="[#1X1:1]-[#6X2:2]-[#1X1:3]"/> 
</Constraints>
```

**Comment by jchodera** [comment]
I guess one question is whether we are OK with the number of atoms tagged in the `smirks` string determine whether to use the equilibrium bond length or whether to interpret the `1-2-3` atom tuple as an angle from which to derive an equilibrium constraint length.

**Comment by davidlmobley** [comment]
I'll ask Chris to look at this. This was sitting there from August, so the comment about asking him in the morning is an old one.

**Comment by cbayly13** [comment]
There are different use cases for constraints, so we either need a superbly designed mother-of-all-API-points or we can go with a fit-for-purpose design for the more immediate needs. IMHO the immediate need is for torsion driving and, conventionally, interaction energies. In these cases we assume the user submits the structure having the desired value of the degree of freedom, so I think it fits in with John's suggestion above: tagged atoms specify the constrained atoms, and they are constrained to their starting value. It would be nice if we could inherit improper torsions, meaning if the tagged 1-2-3-4 atoms are linear, it is a regular torsion, but if it is a trefoil as in 1-2(-3)-4, the whole improper is constrained (all six paths).

**Comment by jchodera** [comment]
I think we might be thinking of different application domains here. This is
just for "which distance constraints are imposed during simulation via
SHAKE or SETTLE?"

OpenMM doesn't currently allow for torsion or other kinda of constraints.
If we do torsion drives, it will be using an external code to impose
geometries.

We don't need to come up with the final set of all possible constraints
either. We just need a syntax that can have stuff added to it later, but
where we don't remove or change things.

**Comment by davidlmobley** [comment]
@jchodera - I just pinged Chris again to clarify that we're addressing constraints for MD simulations, i.e. as normally applied to bonds involving hydrogen or to waters, etc.

In terms of your proposals, here are my thoughts:
- We need an easy mechanism to constrain bond lengths
- We need an easy mechanism to constrain bond angles
- Those two seem like they are separate and can cover all (?) of our typical use cases (?)

So, I like the first part of this example:
```XML
<Constraints>
  <!-- bonds involving hydrogen -->
  <Constraint smirks="[#1:1]-[*:2]"/> <!-- add constraint between atoms 1 and 2 using equilibrium bond distance -->
  <!-- angles involving hydrogen -->
  <Constraint smirks="[#1:1]-[*:2]-[#1:3]"/> <!-- add constraint between atoms 1 and 3 using auto-calculated distance from equilibrium bond and angles -->
</Constraints>
```

At least, I think the "bonds involving hydrogen" case is exactly what we would want a lot. I'm a little nervous about the "angles involving hydrogen" case as it sounds like, in your comment, you would want to interpret this to constrain the 1-3 distance AND the angle. Is that doing too much? (i.e. what if I wanted to constrain all bond angles in a molecule, as is sometimes done, but constrain no bond lengths?)

For example, for water, don't we want to constrain the H-O-H angle and the O-H distance separately, at least in TIP3P? 

To visit your water example:
```XML
<Constraints>
  <!-- constrain water O-H bond to equilibrium bond length -->
  <Constraint smirks="[#1X1:1]-[#6X2:2]-[#1X1]"/> 
  <!-- constrain water H...H, calculating equilibrium length from H-O-H equilibrium angle and H-O equilibrium bond lengths -->
  <Constraint smirks="[#1X1:1]-[#6X2:2]-[#1X1:3]"/> 
</Constraints>
```

That looks right to me, except that I'd be using these SMIRKS strings:
- `[#1:1]-[#8X2H2:2]-[#1]` (oxygen with two connected hydrogens, not carbon with two connections, is the central atom, and the "X1" decorator on the protons is unnecessary)
- `[#1:1]-[#8X2H2:2]-[#1:3]` (same changes as above)
I'd also constrain the bond lengths only based on the first entry and use the second one, which denotes an angle, only to fix the equilibrium angle. If you have the angle case ALSO constrain the bond lengths, then there's no mechanism to apply a constraint only to an angle without also constraining bond lengths, right?

**Comment by jchodera** [comment]
> I'm a little nervous about the "angles involving hydrogen" case as it sounds like, in your comment, you would want to interpret this to constrain the 1-3 distance AND the angle. Is that doing too much? (i.e. what if I wanted to constrain all bond angles in a molecule, as is sometimes done, but constrain no bond lengths?)

We certainly wouldn't want to use this generally, but I was originally proposing this as a way to replace the functionality I stripped out where OpenMM allowed the user to specify `None` (no constraints), `HBonds` (constrain only bonds to hydrogen), and `HAngles` (constrain all bonds and any angles involving hydrogen.

**Comment by jchodera** [comment]
The updated rigid water example is then:
```XML
<Constraints>
  <!-- constrain water O-H bond to equilibrium bond length -->
  <Constraint smirks="[#1:1]-[#8X2H2:2]-[#1]"/> 
  <!-- constrain water H...H, calculating equilibrium length from H-O-H equilibrium angle and H-O equilibrium bond lengths -->
  <Constraint smirks="[#1:1]-[#8X2H2:2]-[#1:3]"/> 
</Constraints>
```

> I'd also constrain the bond lengths only based on the first entry and use the second one, which denotes an angle, only to fix the equilibrium angle. If you have the angle case ALSO constrain the bond lengths, then there's no mechanism to apply a constraint only to an angle without also constraining bond lengths, right?

Oh, I hadn't even thought about that. We can't really constrain an angle specifically in OpenMM without constraining bonds, since only bond constraints are supported, so I'd actually argue we would want `[#1:1]-[#8X2H2:2]-[#1:3]` to constrain three bonds: `(1,2), (2,3), and (1,3)`. Otherwise, we would constrain a distance `(1,3)` but it wouldn't keep the angle fixed.

**Comment by davidlmobley** [comment]
> We certainly wouldn't want to use this generally, but I was originally proposing this as a way to replace the functionality I stripped out where OpenMM allowed the user to specify None (no constraints), HBonds (constrain only bonds to hydrogen), and HAngles (constrain all bonds and any angles involving hydrogen.

It would certainly work for that, though is it preferable to do things this way rather than the slightly more general way of allowing a user to constrain bonds, angles, or both in combination? The latter actually seems easier to understand to me as it doesn't require wrapping my head around the idea that defining an angle SMIRKS also results in bond length constraints.

**Comment by davidlmobley** [comment]
> Oh, I hadn't even thought about that. We can't really constrain an angle specifically in OpenMM without constraining bonds, since only bond constraints are supported, so I'd actually argue we would want [#1:1]-[#8X2H2:2]-[#1:3] to constrain three bonds: (1,2), (2,3), and (1,3). Otherwise, we would constrain a distance (1,3) but it wouldn't keep the angle fixed.

Ah, OK, @jchodera - if there is no mechanism to constrain an angle then I agree with you that the best way to do this is as you're proposing, that an angle constraint results in constraining all both bonds involved in the angle as well as the distance between the terminal atoms.

And I agree with your syntax.

**Comment by jchodera** [comment]
> is it preferable to do things this way rather than the slightly more general way of allowing a user to constrain bonds, angles, or both in combination? The latter actually seems easier to understand to me as it doesn't require wrapping my head around the idea that defining an angle SMIRKS also results in bond length constraints.

I'm trying to think ahead as to what we might want to support in this format. There was initially some discussion about wanting to potentially keep some moieties rigid, or potentially to have groups of atoms (or pseudoatoms) that act like rigid units. This would allow us to have a way of having targeted groups of atoms be rigid in interesting ways, while still making it easy to specify "constraint all bonds to hydrogen" or "constrain all bonds" with a single `<Constraint/>` element.

**Comment by cbayly13** [comment]
Just catching up here, I agree with most of the above discussion about SHAKE and TIP/FastWater. SHAKE/SETTLE is about bond constraints in general, and we certainly need them quite apart from TIP/FastWater. I wonder if FastWater deserves a constraint special case. However, for both SHAKE and FastWater I am quite sure we need to abandon the assumption that the input structures will have the correct values; an obvious case is structured water from the pdb file. For SHAKE and TIP/FastWater we need to specify a constraint value I think.
Also, just FYI, TIP water has an explicit (though artificial) bond from H1 to H2, and thus in SMARTS would be a divalent hydrogen: ```[#1X2]```

**Comment by jchodera** [comment]
What is TIP/FastWater?

SETTLE is the analytical solution of SHAKE for three-site waters, so is specific to three-site molecules.

**Comment by jchodera** [comment]
We do have an example for water that includes this artificial bond from H1 to H2:
```XML
<Constraints>
  <!-- constrain water O-H bond to equilibrium bond length -->
  <Constraint smirks="[#1:1]-[#8X2H2:2]-[#1]"/> 
  <!-- constrain water H1...H2, calculating equilibrium length from H-O-H equilibrium angle and H-O equilibrium bond lengths -->
  <Constraint smirks="[#1:1]-[#8X2H2:2]-[#1:3]"/> 
</Constraints>
```

**Comment by jchodera** [comment]
Any further thoughts, or is this final format OK?

**Comment by davidlmobley** [comment]
I'm good with this, @jchodera . I don't think Chris has any other insights.

**Comment by bannanc** [comment]
I just got finished reading through this, I'm a little confused about the implication of the TIP water with the explicit H-H bond: 

> Also, just FYI, TIP water has an explicit (though artificial) bond from H1 to H2, and thus in SMARTS would be a divalent hydrogen: `[#1X2]`

If we had Tip3p water then the H-H bond would exist and `[#1X2]` would match for hydrogens in water. Assuming you just start from a water molecule then that H-H bond isn't explicitly there and the `[#1X2]` would not match, correct? My understanding is the angle pattern (`[#1:1]-[#8X2H2:2]-[#1:3]`) would then be used to indicate all three distances (not really bond lengths) would be kept constant. 

Otherwise this format makes sense to me.

**Comment by jchodera** [comment]
I don't believe OpenEye tools would be able to correctly process an HOH molecule where the hydrogens were bonded to each other---that would violate its rules of chemistry, from what I understand.

I believe we're thinking of the case where the system contains a bunch of normal H-O-H molecules and we want to apply parameters typically considered to be "TIP3P" parameters. In that case, we would want to impose both H-O and intramolecular H...H constraints, which is why we need a syntax capable of specifying that.

**Comment by davidlmobley** [comment]
I agree with @jchodera here. I don't think Chris was thinking of how we're actually processing molecules/assigning parameters here. He was just thinking that if you had a hypothetical water that was set up like TIP3P, you could have a virtual bond between the two hydrogens and then `[#1X2]` would be a perfectly sensible SMIRKS. But it doesn't make any sense here.

**Comment by davidlmobley** [comment]
@jchodera - we think this is final.

**Comment by jchodera** [comment]
I just spoke with @cbayly13, who pointed out that it would be desirable to be able to express constraint distances as well in this format:
```XML
<Constraints distance_unit="angstroms" angle_unit="degrees">
  <!-- constrain all bonds to hydrogen to their equilibrium bond length -->
  <Constraint smirks="[#1:1]-[*:2]" />
  <!-- constrain water O-H bond to equilibrium bond length (overrides earlier constraint) -->
  <Constraint smirks="[#1:1]-[#8X2H2:2]-[#1]" distance="0.9572"/> 
  <!-- constrain water H...H, calculating equilibrium length from H-O-H equilibrium angle and H-O equilibrium bond lengths -->
  <Constraint smirks="[#1:1]-[#8X2H2]-[#1:2]" distance="1.8532"/> 
  <!-- alternatively, specify the angle, but use equilibrium bond lengths (do we need this?) -->
  <Constraint smirks="[#1:1]-[#8X2H2:2]-[#1:3]" angle="104.52"/> 
</Constraints>
```

**Comment by davidlmobley** [comment]
If I read this correctly, @jchodera , the difference is that you're just specifying distance and angle, potentially? I agree this would be a plus since one might want to be able to start with structures that don't necessarily have correct geometries.

**Comment by davidlmobley** [comment]
Closed, at least temporarily, by #239 , which implements much but not all of this.
